### PR TITLE
Avoid calling ref() of NetworkStorageManager in its constructor

### PR DESCRIPTION
**Pull request by szewai**
#### c05835b72b6159e136f434f891d43321e649affe
<pre>
Avoid calling ref() of NetworkStorageManager in its constructor
<a href="https://bugs.webkit.org/show_bug.cgi?id=254849">https://bugs.webkit.org/show_bug.cgi?id=254849</a>
rdar://107492842

Reviewed by Chris Dumez.

RefCounted::ref() asserts adoption has happened when it is called, so let&apos;s create ThreadSafeWeakPtr instead of Ref in
NetworkStorageManager&apos;s constructor.

* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::activePaths):
(WebKit::NetworkStorageManager::NetworkStorageManager):

Canonical link: <a href="https://commits.webkit.org/262567@main">https://commits.webkit.org/262567@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/796a011d3a318b125117002a4e1af3902f2402f2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1804 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1836 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1894 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2732 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1922 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1908 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1897 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1688 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1822 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1637 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1618 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2572 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1629 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1603 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1527 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1661 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1646 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2735 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1686 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1499 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1611 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1610 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/463 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1756 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp [hunk 1]
```diff
@@ -64,9 +64,9 @@ static const Seconds defaultBackupExclusionPeriod { 24_h };
 #endif
 
 // FIXME: Remove this if rdar://104754030 is fixed.
-static HashMap<String, WeakPtr<NetworkStorageManager>>& activePaths()
+static HashMap<String, NetworkStorageManager*>& activePaths()
```
Comment (cdumez) [inline Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp]: Why this change? We try to move away from raw pointers nowadays.

File: Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp [hunk 2]
```diff
@@ -132,63 +132,75 @@ static void deleteEmptyOriginDirectory(const String& directory)
...
-    , m_defaultOriginQuota(defaultOriginQuota)
-    , m_defaultThirdPartyOriginQuota(defaultThirdPartyOriginQuota)
-    , m_originQuotaRatio(originQuotaRatio)
-    , m_totalQuotaRatio(totalQuotaRatio)
-    , m_volumeCapacityOverride(volumeCapacityOverride)
-    , m_parentConnection(connection)
+void NetworkStorageManager::initialize(const String& path, const String& customLocalStoragePath, const String& customIDBStoragePath, const String& customCacheStoragePath, uint64_t defaultOriginQuota, uint64_t defaultThirdPartyOriginQuota, std::optional<double> originQuotaRatio, std::optional<double> totalQuotaRatio, std::optional<uint64_t> volumeCapacityOverride, UnifiedOriginStorageLevel level)
+{
+    m_queue->dispatch([this, protectedThis = Ref { *this }, path = path.isolatedCopy(), customLocalStoragePath = customLocalStoragePath.isolatedCopy(), customIDBStoragePath = customIDBStoragePath.isolatedCopy(), customCacheStoragePath = customCacheStoragePath.isolatedCopy(), defaultOriginQuota, defaultThirdPartyOriginQuota, originQuotaRatio, totalQuotaRatio, volumeCapacityOverride, level]() mutable {
```
Comment (cdumez) [inline Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp]: Could we have captured a ThreadSafeWeakPtr instead and avoided the refactoring?

Comment (szewai) [reply Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp]: Sure, I am a little confused by all these pointers.
Is it safe to create a ThreadSafeWeakPtr inside constructor?

Comment (cdumez) [reply Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp]: Yes, it is safe to create WeakPtr or TheadSafeWeakPtr from inside constructors.

Also note that it is generally safe to ref on object from inside its constructor as far as I know. However, you'd need to relax adoption requirements (there is a function could can call for that) to avoid hitting the assertion in debug. The point of the assertion wasn't to stop you from ref'ing the object from inside the constructor. Its point is to make sure that people don't forget to adopt newly constructed refcounted objects.

Anyway, here, my recommendation would be to use ThreadSafeWeakPtr<> since I don't think we truly need to extend the lifetime and since it looks a little cleaner than relaxing adoption requirements.

File: Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp [hunk 3]
```diff
@@ -139,29 +139,35 @@ NetworkStorageManager::NetworkStorageManager(PAL::SessionID sessionID, Markable<
     : m_sessionID(sessionID)
     , m_queueName(makeString("com.apple.WebKit.Storage.", sessionID.toUInt64(), ".", static_cast<uint64_t>(identifier->data() >> 64), static_cast<uint64_t>(identifier->data())))
     , m_queue(SuspendableWorkQueue::create(m_queueName.utf8().data(), SuspendableWorkQueue::QOS::Default, SuspendableWorkQueue::ShouldLog::Yes))
-    , m_defaultOriginQuota(defaultOriginQuota)
```
Comment (cdumez) [inline Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:142]: Why did you make these initializer changes?

Comment (szewai) [reply Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:142]: They are supposed to be accessed from the WorkQueue only

File: Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp [hunk 4]
```diff
@@ -139,29 +139,35 @@ NetworkStorageManager::NetworkStorageManager(PAL::SessionID sessionID, Markable<
...
+
 {
     ASSERT(RunLoop::isMain());
 
     if (!path.isEmpty()) {
         auto addResult = activePaths().add(path, *this);
         if (!addResult.isNewEntry) {
-            if (auto existingManager = addResult.iterator->value)
+            auto existingManager = addResult.iterator->value.get();
+            if (existingManager)
```
Comment (cdumez) [inline Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp]: No reason to increase scoping here it seems? Why didn't you use:
```
if (auto existingManager = addResult.iterator->value.get())
```

File: Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp (hunk 3 above)
Comment (cdumez) [reply Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:142]: I hope that's true we'll get new weird behavior :)

**Review by cdumez — APPROVED** [review]
